unit/goto-programs/remove_virtual_functions_without_fallback.cpp

```diff
@@ -0,0 +1,309 @@
+
+ Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+#include <testing-utils/load_java_class.h>
+
+#include <util/simplify_expr.h>
+#include <goto-programs/remove_virtual_functions.h>
+
+/// Try to resolve a classid comparison `expr`, assuming that any accessed
+/// classid has value `actual_class_id`
+exprt resolve_classid_test(
+  const exprt &expr,
+  const irep_idt &actual_class_id,
+  const namespacet &ns)
+{
+  if(expr.id() == ID_or || expr.id() == ID_and)
+  {
+    exprt resolved = expr;
+    for(exprt &op : resolved.operands())
+      op = resolve_classid_test(op, actual_class_id, ns);
+    simplify(resolved, ns);
+    return resolved;
+  }
+
+  if(expr.op0().id() == ID_constant && expr.op1().id() != ID_constant)
+  {
+    exprt swapped = expr;
+    std::swap(swapped.op0(), swapped.op1());
+    return resolve_classid_test(swapped, actual_class_id, ns);
+  }
+
+  if(expr.op0().id() == ID_member &&
+     to_member_expr(expr.op0()).get_component_name() == "@class_identifier" &&
+     expr.op1().id() == ID_constant &&
+     expr.op1().type().id() == ID_string)
+  {
+    exprt resolved = expr;
+    resolved.op0() = constant_exprt(actual_class_id, expr.op1().type());
+    simplify(resolved, ns);
+    return resolved;
+  }
+
+  return expr;
+}
+
+static bool is_call_to(
+  goto_programt::const_targett inst, const irep_idt &callee)
+{
+  if(!inst->is_function_call())
+    return false;
+  const exprt &callee_expr = to_code_function_call(inst->code).function();
+  if(callee_expr.id() != ID_symbol)
+    return false;
+  return to_symbol_expr(callee_expr).get_identifier() == callee;
+}
+
+static bool is_assume_false(goto_programt::const_targett inst)
+{
+  return inst->is_assume() && inst->guard.is_false();
+}
+
+/// Interpret `program`, resolving classid comparisons assuming any actual
+/// callee has class `actual_class_id`, returning the first instruction we
+/// arrive at which is neither a SKIP, nor a GOTO that can be resolved in this
+/// way.
+static goto_programt::const_targett interpret_classid_comparison(
+  const goto_programt &program,
+  const irep_idt &actual_class_id,
+  const namespacet &ns)
+{
+  REQUIRE(!program.instructions.empty());
+  goto_programt::const_targett pc = program.instructions.begin();
+
+  while(pc != program.instructions.end())
+  {
+    if(pc->type == GOTO)
+    {
+      exprt guard = pc->guard;
+      guard = resolve_classid_test(guard, actual_class_id, ns);
+      if(guard.is_true())
+      {
+        REQUIRE(pc->targets.begin() != pc->targets.end());
+        pc = *(pc->targets.begin());
+      }
+      else if(guard.is_false())
+        ++pc;
+      else
+      {
+        // Can't resolve the test, so exit here:
+        return pc;
+      }
+    }
+    else if(pc->type == SKIP)
+    {
+      ++pc;
+    }
+    else
+    {
+      return pc;
+    }
+  }
+
+  return program.get_end_function();
+}
+
```

Comment:
Would it make sense to extract the above functions as utility functions?

Reply:
I don't think so, they're really quite specific in purpose and the classid test substitution thing is not generally sound